Improve README a bit more (#282)
kowainik/relude: README.md

[![Stackage Nightly](http://stackage.org/package/relude/badge/nightly)](http://stackage.org/nightly/package/relude)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

**`relude`** is a safe, performant, user-friendly and lightweight Haskell
standard library.

The default `Prelude` is not perfect and doesn't always satisfies one's needs.
At this stage you may want to try an alternative prelude library. In this README
we are going to give you convincing reasons to consider using `relude` as such
alternative in your next project.

`relude` has some strong goals and principles that it sticks to. That principles
defines the library decisions and might tell you more about the priorities of
the library. So below you can find the key design principles behind `relude`:

1. **Productivity.** You can be more productive with a "non-standard" standard

>    library.

Reviewer: Excellent idea to put **Productivity as the main design principle! 👏

2. **Avoid all** [**partial functions**](https://www.reddit.com/r/haskell/comments/5n51u3/why_are_partial_functions_as_in_head_tail_bad/)
   (e.g. `head :: [a] -> a`). The types of partial functions lie about their
   behavior and usage of such functions can lead to unexpected bugs and runtime
   exceptions in pure code.

> 3. **Type-safety**. We use *"make invalid states unrepresentable"* motto as one

Reviewer: ```suggestion
3. **Type-safety**. We use the *"make invalid states unrepresentable"* motto as one
```

   of our principles. If it is possible, we express this concept through the
   types.

    _Example:_ Here the function's name type and actions are aligned with each other

    ```haskell
    whenNotNull :: Applicative f => [a] -> (NonEmpty a -> f ()) -> f ()
    ```
    While in here, the type could represent an unreachable state:

    ```haskell
    whenNotNull :: Applicative f => [a] -> ([a] -> f ()) -> f ()
    ```
4. **Performance.** We prefer `Text` over [`String`](https://www.reddit.com/r/haskell/comments/29jw0s/whats_wrong_with_string/),

>    use space-leaks-free functions (e.g. our custom performant `sum` and `product`), introduce

Reviewer: I like new rewording of this paragraph 😊

   `{-# INLINE #-}` and `{-# SPECIALIZE #-}` pragmas where appropriate.
5. **Minimalism** (low number of dependencies). We don not force users of
   `relude` to stick to any specific lens or text formatting or logging
   library. Where possible, `relude` depends only on boot libraries.
   The [Dependency graph](https://raw.githubusercontent.com/kowainik/relude/master/relude-dependency-graph.png)
   of `relude` can give you a clearer picture.
6. **Convenience** (e.g. lifted to `MonadIO` functions, more reexports). Despite
   minimalism, we want to bring common types and functions (like `containers`
   and `bytestring`) into scope because they are used in almost every
   application anyways.
7. **Provide excellent documentation.**
   + Tutorial
   + Migration guide from `Prelude`
   + Haddock for every function with examples tested by
     [`doctest`](http://hackage.haskell.org/package/doctest)
   + Documentation on [internal module structure](http://hackage.haskell.org/package/relude/docs/Relude.html)
   + `relude`-specific [HLint](http://hackage.haskell.org/package/hlint) rules:
     [`.hlint.yaml`](https://github.com/kowainik/relude/blob/master/.hlint.yaml)

> 8. **User-friendliness.** Anyone is able to quickly migrate to `relude`. Only

Reviewer: Is it better to change wording here, so it will sound more like a goal?
```suggestion
8. **User-friendliness.** Anyone should be able to quickly migrate to `relude`. Only
```

   some basic familiarity with the common libraries like `text` and `containers`
   should be helpful (but not necessary).
9. **Exploration.** We have space to experiment with new ideas and proposals
   without introducing breaking changes. `relude` uses the approach with
   `Extra.*` modules which are not exported by default, so it makes it quite
   easy for us to provide new functionality without breaking anything and let
   the users decide to use it or not.

> <!-- TODO: anti-goals? -->

Reviewer: Yes, I think it will be an excellent addition to the README to tell our anti-goals as well 👍

2. Extra dependencies: adding more libraries to dependencies increases build
   times and maintenance burden.
   + `relude` depends only on the boot libraries (almost) which results in small
     build time, follows [PVP](https://pvp.haskell.org/) and cares about
     backwards compatibility.

However, using an alternative prelude, specifically `relude`, has the following
**advantages**:

1. Increased code safety: no partial functions, no space-leak functions.
2. Increased productivity: no need to import common functions and data types,
   more common idioms are provided.
3. Increased performance: some functions in `relude` are faster than in the
   default `Prelude`.

Taking into consideration all the above points, we put together our
recommendations when to use `relude`:

1. When you develop an application (e.g. CLI tool, web-app). In that case,
   greater productivity is more important than a low number of dependencies.
2. When writing a big framework. Some of them can be bigger than applications.

And when you may want to stay with the default standard:

1. When you write a small library that is supposed to be used by other people in

>    their projects.

Reviewer: This is a really nice idea to write a reason for not using a `relude` 👍

  default-language:    Haskell2010
```

> **NOTE:** if you use [`summoner`](https://github.com/kowainik/summoner) to

> > generate a Haskell project, this tool can automatically create the `mixins`

Reviewer: ```suggestion
> generate a Haskell project, the automatically creates the `mixins`
```

> > field when you specify the custom prelude.

Reviewer: ```suggestion
> field when you specify a custom prelude.
```

  + `show` is polymorphic over return type.
  + Functions `sum` and `product` are strict now, which makes them more efficient.
  + You can't call `elem` and `notElem` functions over `Set` and `HashSet`.

>     These functions are forbidden for these two types for performance reasons.

Reviewer: ```suggestion
    These functions are forbidden for these two types due to performance reasons.
```

### Deepseq

All the main parts of the `deepseq` library are exported.
For instance, if you want to force the deep evaluation of
some value (in `IO`), you can write `evaluateNF a`.
Additionally, the WHNF evaluation is possible
with provided `evaluateWHNF`.

## What's new?

[[Back to the Table of Contents] ↑](#structure-of-this-tutorial)

Finally, let's move to part describing the new cool features we bring with
`relude`.

### Available by default

* Safe analogue for the list functions:
  ```haskell

>   head :: NonEmpty a -> a

Reviewer: I thought it would be cool to specify all such functions in here.
```suggestion
  head :: NonEmpty a -> a    -- the first element of the list
  tail :: NonEmpty a -> [a]  -- the list without the first element
  last :: NonEmpty a -> a    -- the last element of the list
  init :: NonEmpty a -> [a]  -- the list without the last element
```

      f x
  ```

* `for_` and `forM_` for loops.

  ```haskell

>   for_ [1..10] $ \i -> do

Reviewer: I just realized that `for_ [1..10]` is actually an antipattern, so let's replace indexed for-loop with foreach-loop...
```suggestion
  for_ files $ \file -> do
```

  non-empty structures like `NonEmpty`.
* [`StaticMap` and `DynamicMap` type classes](src/Relude/Extra/Map.hs) as a
  general interface for `Map`-like data structures.
* And much more!

Explore `Extra` modules: [`Relude.Extra`](http://hackage.haskell.org/package/relude/docs/Relude-Extra.html)

## Migration guide

[[Back to the Table of Contents] ↑](#structure-of-this-tutorial)

This section of the guide helps to migrate your project from `base`
to the `relude` library.

In order to replace the default `Prelude` with `relude` you should start with

> instructions provided in the  [_Get Started_](#get-started-) section.

Author: ```suggestion
instructions provided in the  [_Get Started_](#get-started) section.
```
